### PR TITLE
Feature/depict - Part 2 (Reactions + Layouting)

### PR DESCRIPTION
**Pull request by johnmay**
Part 2 - mainly tweaks to depiction classes (now immutable), reaction layout, and some long overdue patches to the SDG (allow it to handle fragments).

Rendering code hard to test as usual but added some for the layout updates.

#### Reaction Layout

Testing the removal of parts, here's the original...

```
CNC1=C(OCCC#N)C=CC=C1.CCCC[Sn](CCCC)(CCCC)N=[N+]=[N-].CC[Al]CC)CC>CC1=CC=CC=C1.CCOC(C)=O.O>CNC1=C(OCCC2=NN=NN2)C=CC=C1
```

No Agents...

Products Only...

Reactants Only...

Empty Reaction - (failed reaction)...

Agents only... it doesn't depict the arrow, actually I think that makes sense!

#### Atom Maps and Highlight

In the generator we turn on `withAtomMapNumbers()` or `withAtomMapHighlight()` to show the atom-mapping between reactants and products (supplied in the input SMILES).

```
CC1(C)CCCC(C)(C)N1.C([Li])CCC.[s:16]1[cH:20][cH:19][cH:18][c:17]1-[c:21]1[c:29]2[c:25]([n:26][s:27][n:28]2)[c:24](-[c:30]2[s:31][cH:32][cH:33][cH:34]2)[cH:23][cH:22]1.[CH3:35][Sn:36](Cl)([CH3:38])[CH3:37]>C1OCCC1.C(OCC)C.O>[CH3:35][Sn:36]([CH3:38])([CH3:37])[c:32]1[s:31][c:30](-[c:24]2[c:25]3[c:29]([n:28][s:27][n:26]3)[c:21](-[c:17]3[s:16][c:20]([Sn:36]([CH3:38])([CH3:37])[CH3:35])[cH:19][cH:18]3)[cH:22][cH:23]2)[cH:34][cH:33]1
```

```
Cl[c:2]1[c:11]2[c:6]([c:7]3[cH:15][cH:14][cH:13][cH:12][c:8]3[cH:9][cH:10]2)[n:5][cH:4][n:3]1.O1CCCC1.[Br-].[CH2:22]([Mg+])[CH:23]([CH3:25])[CH3:24].Cl>CN1CCCC1=O>[CH2:22]([c:2]1[c:11]2[c:6]([c:7]3[cH:15][cH:14][cH:13][cH:12][c:8]3[cH:9][cH:10]2)[n:5][cH:4][n:3]1)[CH:23]([CH3:25])[CH3:24]
```

#### Salt Layout

By default, the SDG now lays out fragments of a molecule in a grid. If possible it tries to (re)bond ionic fragments before doing so with temporary bonds. Here's some results:

Sodium Benzoate

image

AlCl3
`[Al+3].[Cl-].[Cl-].[Cl-]`

image

K2CO3
`[K+].[O-]C=O[O-][K+]`

CHEMBL11933 - avoids the nitro

### CODE REVIEW

**Comment by egonw** [comment]
The images are not working for me...

File: app/depict/src/main/java/org/openscience/cdk/depict/DepictionGenerator.java [hunk 1]
```diff
@@ -98,6 +101,40 @@
 public final class DepictionGenerator {
 
     /**
+     * Visually distinct colors for highlighting.
+     * http://stackoverflow.com/a/4382138
+     * Kenneth L. Kelly and Deanne B. Judd.
+     * "Color: Universal Language and Dictionary of Names",
+     * National Bureau of Standards,
+     * Spec. Publ. 440, Dec. 1976, 189 pages.
```
Comment (egonw) [inline app/depict/src/main/java/org/openscience/cdk/depict/DepictionGenerator.java]: Would be nice to put this in cheminf.bibx too and user {@cdk.cite ...}

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java [hunk 2]
```diff
@@ -287,6 +293,19 @@ public void generateExperimentalCoordinates(Vector2d firstBondVector) throws CDK
...
+        generateCoordinates(firstBondVector, false);
+    }
+
+    /**
+     *  The main method of this StructurDiagramGenerator. Assign a molecule to the
+     *  StructurDiagramGenerator, call the generateCoordinates() method and get
+     *  your molecule back.
+     *
+     *  @param  firstBondVector          The vector of the first bond to lay out
+     *  @throws CDKException             if an error occurs
```
Comment (egonw) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: Please add a @param for isFragment.

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java [hunk 3]
```diff
@@ -297,12 +316,23 @@ public void generateCoordinates(Vector2d firstBondVector) throws CDKException {
...
+            return;
         }
-        if (!ConnectivityChecker.isConnected(molecule)) {
-            logger.debug("Molecule is not connected. Throwing exception.");
-            throw new CDKException(disconnectedMessage);
-        } else {
-            logger.debug("Molecule is connected.");
+
+        // intercept fragment molecules and lay them out in a grid
+        if (!isFragment) {
```
Comment (egonw) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: I don't get the combo of the above two lines... if it is a isFragment, then it should do the putting on the grid, not?

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java [hunk 4]
```diff
@@ -465,12 +495,297 @@ public void generateCoordinates(Vector2d firstBondVector) throws CDKException {
...
+     * of two or more fragments and creates zero or more bonds (return value) that
+     * should be temporarily used for layout generation. In general this problem is
+     * difficult but since molecules will be laid out in a grid by default - any
+     * positioning is an improvement. Heuristics could be added if bad (re)bonds
+     * are seen.
+     *
+     * @param frags connected fragments
+     * @return ionic bonds to make
+     */
+    private List<IBond> makeIonicBonds(final List<IAtomContainer> frags) {
```
Comment (egonw) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: Great work!

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java [hunk 5]
```diff
@@ -581,18 +896,18 @@ private void layoutRingSet(Vector2d firstBondVector, IRingSet rs) throws CDKExce
          * this.mappedSubstructures
          */
         if (useTemplates && mappedSubstructures.getAtomContainerCount() > 0
-                && System.getProperty("java.version").indexOf("1.3.") == -1) {
+            && System.getProperty("java.version").indexOf("1.3.") == -1) {
```
Comment (egonw) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: Do you have any idea why this was here? Is it still relevant? Do you think we still have people using Java 1.3??

**Comment by johnmay** [comment]
Doh images aren't shareable from google. Odd because the last ones were

**Comment by johnmay** [comment]
Hopefully visible now but I think its google moving links.. probably to prevent hot linking.

Here's the album link - https://goo.gl/photos/J99araookYzmTsa6A

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java (hunk 3 above)
Comment (johnmay) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: Ah bad naming.. param could be named better 'isConnected'.. a fragment is connected... fragments are not..

File: tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java (hunk 5 above)
Comment (johnmay) [inline tool/sdg/src/main/java/org/openscience/cdk/layout/StructureDiagramGenerator.java]: Nope... I do remember a bug report somewhere.

**Pull request merged**